sonarqube/resource_sonarqube_portfolio.go

```diff
@@ -89,30 +106,56 @@ func resourceSonarqubePortfolio() *schema.Resource {
 				Type:          schema.TypeString,
 				Optional:      true,
 				ForceNew:      false,
-				ConflictsWith: []string{"tags"},
+				ConflictsWith: []string{"selected_projects", "tags"},
 				ValidateFunc:  validation.StringIsValidRegExp,
 			},
-			// TODO: MANUAL
-			// "selectedProjects": [],
-			// "projects": [],
+			"selected_projects": {
+				Type:          schema.TypeSet,
+				Optional:      true,
+				ForceNew:      false,
+				ConflictsWith: []string{"tags", "regexp"},
+				Description:   "A set of projects to add to the portfolio.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"project_key": {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "The project key of the project to add to the portfolio",
+						},
+						"selected_branches": {
+							Type:        schema.TypeSet,
+							Optional:    true,
+							Description: "A set of branches for the project to add to the portfolio",
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+					},
+				},
+			},
+			"projects": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
 		},
 	}
 }
 
 func checkPortfolioSupport(conf *ProviderConfiguration) error {
 	if strings.ToLower(conf.sonarQubeEdition) != "enterprise" {
-		return fmt.Errorf("Portfolios are only supported in the Enterprise edition of SonarQube. You are using: SonarQube %s version %s", conf.sonarQubeEdition, conf.sonarQubeVersion)
+		return fmt.Errorf("portfolios are only supported in the Enterprise edition of SonarQube. You are using: SonarQube %s version %s", conf.sonarQubeEdition, conf.sonarQubeVersion)
 	}
 	return nil
 }
 
 // Validate the regexp and tag fields if the corresponding selection_mode is chosen
 func validatePortfolioResource(d *schema.ResourceData) error {
 	switch selectionMode := d.Get("selection_mode"); selectionMode {
-	case "NONE", "MANUAL", "REST":
+	// TODO: Validate MANUAL properly
```

Comment:
Ops I forgot to remove the comment. My idea when I wrote it was that if MANUAL mode was selected, then there should be a validation that selected_projects should have at least one value in it, but I later realized that should not be done.

Firstly because you can just create a MANUAL portfolio and not populate it in the UI. The same should be possible in the provider. But also, if anyone today is using the provider to define MANUAL portfolios, and then add projects to it manually, then using the provider after this goes through would crash.

Reply:
Actually, that's a test case I didn't cover. What will happen to MANUAL portfolios with projects added outside of terraform. I think the behaviour should be that a bump in version of this provider should not change them at all, but I'll have to test it. I'll do that today or tomorrow